src/Model/LazyBSONDocument.php

```diff
@@ -0,0 +1,316 @@
+
+use function array_filter;
+use function array_key_exists;
+use function array_map;
+use function count;
+use function get_object_vars;
+use function is_array;
+use function is_object;
+use function is_string;
+use function iterator_to_array;
+use function MongoDB\recursive_copy;
+use function sprintf;
+use function trigger_error;
+
+use const E_USER_WARNING;
+
+/**
+ * Model class for a BSON document.
+ *
+ * The internal data will be cast to an object during BSON serialization to
+ * ensure that it becomes a BSON document.
+ *
+ * @template TValue
+ * @template-implements ArrayAccess<string, TValue>
+ * @template-implements IteratorAggregate<string, TValue>
+ */
+final class LazyBSONDocument implements ArrayAccess, Countable, IteratorAggregate, JsonSerializable
+{
+    /** @var Document<TValue> */
+    private Document $bson;
+
+    /** @var array<string, TValue> */
+    private array $read = [];
+
+    /** @var array<string, bool> */
+    private array $exists = [];
+
+    /** @var array<string, TValue> */
+    private array $set = [];
+
+    /** @var array<string, true> */
+    private array $unset = [];
+
+    private bool $entireDocumentRead = false;
+
+    private CodecLibrary $codecLibrary;
+
+    /**
+     * Deep clone this lazy document.
+     */
+    public function __clone()
+    {
+        $this->bson = clone $this->bson;
+
+        foreach ($this->set as $key => $value) {
+            $this->set[$key] = recursive_copy($value);
+        }
+    }
+
+    /** @param Document<TValue>|array<string, TValue>|object|null $input */
+    public function __construct($input = null, ?CodecLibrary $codecLibrary = null)
+    {
+        if ($input === null) {
+            $this->bson = Document::fromPHP([]);
+        } elseif ($input instanceof Document) {
+            $this->bson = $input;
+        } elseif (is_array($input) || is_object($input)) {
+            $this->bson = Document::fromPHP([]);
+
+            if (is_object($input)) {
+                $input = get_object_vars($input);
+            }
+
+            foreach ($input as $key => $value) {
+                $this->set[$key] = $value;
+                $this->exists[$key] = true;
+            }
+        } else {
+            throw InvalidArgumentException::invalidType('input', $input, [Document::class, 'array', 'null']);
+        }
+
+        $this->codecLibrary = $codecLibrary ?? new LazyBSONCodecLibrary();
+    }
+
+    /** @return TValue */
+    public function __get(string $property)
+    {
+        $this->readFromBson($property);
+
+        if (isset($this->unset[$property]) || ! $this->exists[$property]) {
+            trigger_error(sprintf('Undefined property: %s', $property), E_USER_WARNING);
+
+            return null;
+        }
+
+        return array_key_exists($property, $this->set) ? $this->set[$property] : $this->read[$property];
+    }
+
+    public function __isset(string $name): bool
+    {
+        return $this->exists[$name] ??= $this->bson->has($name);
```

Comment:
I can revisit this once benchmarking is in place. At the moment, `$this->read` expects to contain deserialised items, so if we were to read this we'd be obliged to also decode it, which would present a significant performance penalty (along with a memory penalty). However, just storing the raw value in case the property is indeed accessed again may be worth not having to iterate the structure again when a user calls `get` after checking whether the property exists (which is going to be a common use case I imagine).

Reply:
I'll defer to you to jot this down or create some ticket dependent on benchmarks being implemented. Feel free to resolve this thread on your own.

Reply:
Created PHPLIB-1213.